Add vimeo/psalm

lukaslangen opened the pull request:
Solves issue #73 

- [x] Create a .psalm.xml.dist in the project root
- [x] Copy and paste the contents from this psalm.xml.dist
- [x] Run $ composer require vimeo/psalm
- [x] Run $ vendor/bin/psalm --set-baseline=psalm-baseline.xml
- [x] Add a composer script static-analysis with the command psalm --shepherd --stats
- [x] Add a new line to script: in .travis.yml: - if [[ $TEST_COVERAGE == 'true' ]]; then composer static-analysis ; fi
- [x] **(obsolete, because no phpstan configuration was given)** _Remove phpstan from the project (phpstan.neon.dist, .travis.yml entry, composer.json require-dev and scripts)_

Commit:
Add vimeo/psalm

Added configuration in .travis.yml and composer.json
Added psalm-baseline.xml

Signed-off-by: Lukas Langen <lukas@langen.dev>

boesing commented:
@lukaslangen Thanks for this PR. 

To make travis pass, you can either temporarily add `COMPOSER_ROOT_VERSION=2.14.0` to `travis.yml` or `"version": "2.14.0"` to `composer.json` in a single commit.
After this, travis should startup and execute tests.
When anything is green, we can add this to the next release 👍

lukaslangen committed:
Add version for travis

Signed-off-by: Lukas Langen <lukas@langen.dev>

weierophinney committed:
qa: move CS checks to PHP 7.3

Due to issues running on 7.4

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: fix annotation for property

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
@lukaslangen Thank you!

I've rebased and pushed back to your branch. Since the 2.14.x series has a number of other changes, I'm resetting the baseline, and correcting a number of the autogenerated `@psalm-return` annotations on data providers; once those are done, I'll push back here, and it should be ready to merge.

Thanks for getting this started!

weierophinney committed:
qa: create new Psalm baseline and apply low-hanging fruit

- Updates to Psalm 4.3 series and psalm/plugin-phpunit 0.15 series, to correct issues when defining array/iterable annotations.
- Applies autofixes.
- Adds appropriate `@psalm-return` annotations for unit test data providers.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius commented on src/Barcode.php:
```diff
@@ -118,18 +118,19 @@ public function setAdapter($adapter, $options = null)
+     * @return bool
      */
-    public function getChecksum()
+    public function getChecksum(): bool
```
Thos change seems wrong 🤔

weierophinney replied:
Yeah, ran across another in the Bitwise validator. The original patch was on an older version of Psalm, so when I did my autofixes, it wasn't taking into account BC rules (which the v4 series does). Doing an audit of signatures now.

weierophinney committed:
qa: fix return type on Bitwise validator

Psalm added `: int`, which (a) breaks BC, and (b) is incorrect, as it can sometimes return a null.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: fix BC breaks

Since I did autocorrections originally with an older version of Psalm, it did not take into account backwards compatibility, and introduced a number of BC breaks.
This patch corrects them.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: fix data provider name

Use camelCase, not TitleCase.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: update baseline

One more error that was invalid to add to the initial package baseline

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
